content/macro-annotations.md

```diff
@@ -0,0 +1,349 @@
+
+
+
+### Compatibility
+The introduction of the `MacroAnnotation` trait does not affect binary nor TASTy compatibility.
+
+Crucially, macro annotations cannot break TASTy compatibility because they are evaluated after pickling.
+They should also not break binary compatibility as newly added definitions by macro annotations are assumed to only be used from within the macro expanded code.
+
+Macro annotations can be used to patch binary incompatibilities by generating an old definition that is not generated anymore.
+
+
+### Feature Interactions
+
+
+#### Typer
+As we transform the definitions after pickling, the typer only needs to be able to type the reference to the macro annotation.
+
+#### Separate/incremental compilation
+As we transform the definitions after pickling, separate/incremental compilation is not affected by the macro annotations.
+
+#### Compilation suspension
+Inline macros introduced the concept of compilation suspension to be able to define and use a macro in the same project.
+If a call to a macro discovers statically or dynamically that it cannot expand because the macro implementation has not been compiled yet, then the compiler will suspend the compilation of the file that contains the macro call.
+Once it finishes the compilation of other files, the compiler will retry the compilation of the suspended file.
+This is repeated until all files are compiled or no progress is made due to a cyclic macro dependency.
+
+This same idea applies to macro annotations. The macro annotation `transform` method must be compiled before we can expand the macro annotation in another file.
+
+#### Inlining and macros
+Given that macro annotations can generate code that contains inline calls and that inline code can contain macro annotations,
+we need to ensure that after any expansion of a macro annotation or inline call we check for further macro expansions.
+This is already the case with inline call, we just need to make sure the same happens for macro annotations.
+
+Transparent inline method calls generated by macro annotations will expand in the same way as a transparent inline method call inside a non-transparent inline method definition.
+This implies that the type refinement of the transparent inline method will be limited and will not affect the typing process.
+
+
+#### IDE and other tools
+
+Given that we only generate code after pickling, the IDEs that consume TASTy will be agnostic to these changes. All
+definitions will align with source and typer as it does for non-annotated classes.
+
+
+## Alternatives
+
+### Change the order of evaluation of nested macro annotations
+There might be reasons to evaluate the inner macro annotation first, then the outer one.
+In that case, we have two options: we reverse the order of evaluation, or we add the ability to specify the order of evaluation of each macro annotation.
+
+### Allow new definitions to be macro annotated
```

Comment:
I don't understand what this means.

Reply:
This refers to new definitions created by the macro expansion.